### ISSUE TITLE
Detect Objects Exposed via javax.rmi.ssl.SslRMIClientSocketFactory

### ISSUE THREAD

**Issue by NickstaDB** [BishopFox/rmiscout#4]
Eh up! Nice work, love a bit of RMI hackery! I had a chance to test this out on a client engagement recently and hit a gotcha that might be worth handling and documenting.

Essentially an RMI service can associate a socket factory with a bound object reference. Method invocations against the returned object reference will use the socket factory class requested by the server (see https://docs.oracle.com/javase/8/docs/technotes/guides/rmi/socketfactory/index.html#usingCustomSocketFactory) to create a socket for the connection and method invocation. In this case `javax.rmi.ssl.SslRMIClientSocketFactory` was used along with a self-signed certificate. This resulted in a `java.rmi.ConnectIOException` being thrown here https://github.com/BishopFox/rmiscout/blob/master/src/main/java/com/bishopfox/rmiscout/RMIConnector.java#L313. The full stack trace is included below for reference.

It seems to be a PitA to get Java to accept self-signed certs but the solution I came up with was:

1. Capture the network traffic from `rmiscout` to identify the message from the server containing the TLS certificate
2. Save the certificate to a file (in my case with Wireshark, select the `Certificate: xxxx` row in the packet details pane, right-click and select `Export Packet Bytes`)
3. Import the certificate into the Java keystore (on Ubuntu 18 I did this with `keytool -import -keystore /etc/ssl/certs/java/cacerts -storepass changeit -file cert.crt`)

After importing the certificate, `rmiscout` was able to attempt method invocations on the object. Sadly no luck this time :(

You can probably hack something together to handle this cleanly (i.e. possibly altering the socket factory, not sure if proxying, similar to what I do with BaRMIe, would be required), but it's probably worth at least catching the exception and bailing out so `rmiscout` doesn't attempt 100s or 1000s of connections to a service it can't speak to.

```
java.rmi.ConnectIOException: error during JRMP connection establishment; nested exception is: 
	javax.net.ssl.SSLHandshakeException: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.rmi.transport.tcp.TCPChannel.createConnection(TCPChannel.java:307)
	at sun.rmi.transport.tcp.TCPChannel.newConnection(TCPChannel.java:202)
	at sun.rmi.server.UnicastRef.invoke(UnicastRef.java:132)
	at com.bishopfox.rmiscout.RMIConnector.execute(RMIConnector.java:313)
	at com.bishopfox.rmiscout.RMIConnector.checkIfPresent(RMIConnector.java:241)
	at com.bishopfox.rmiscout.RMIScout.process(RMIScout.java:197)
	at com.bishopfox.rmiscout.RMIScout.main(RMIScout.java:134)
Caused by: javax.net.ssl.SSLHandshakeException: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:198)
	at sun.security.ssl.SSLSocketImpl.fatal(SSLSocketImpl.java:1967)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:331)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:325)
	at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1688)
	at sun.security.ssl.ClientHandshaker.processMessage(ClientHandshaker.java:226)
	at sun.security.ssl.Handshaker.processLoop(Handshaker.java:1082)
	at sun.security.ssl.Handshaker.process_record(Handshaker.java:1010)
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1079)
	at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1388)
	at sun.security.ssl.SSLSocketImpl.writeRecord(SSLSocketImpl.java:765)
	at sun.security.ssl.AppOutputStream.write(AppOutputStream.java:123)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at sun.rmi.transport.tcp.TCPChannel.createConnection(TCPChannel.java:229)
	... 6 more
Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:450)
	at sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:317)
	at sun.security.validator.Validator.validate(Validator.java:262)
	at sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:330)
	at sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:237)
	at sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:132)
	at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1670)
	... 17 more
Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:141)
	at sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:126)
	at java.security.cert.CertPathBuilder.build(CertPathBuilder.java:280)
	at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:445)
	... 23 more
```

**Comment by the-bumble** [comment]
Hey @NickstaDB!

Thank you for the bug report! Also, thank you for your work with BarMIE and your 44con presentation. Both were an inspiration and helpful when I was building RMIScout :)

This issue should be fixed by 4e8d985bf9d4c9b66a8478105a265716db3c382e and the v1.02 release. Created a custom `RMIClientSocketFactory` with an all-trusting `TrustManager`. Given that we're just bruteforcing or running exploits, I decided to  spare users from dealing with the Java keystore/validation XD.

Cleartext connections will be attempted first, and then it will automatically attempt a permissive SSL connection before bailing out. 

Happy hacking!

Jake

**Comment by NickstaDB** [comment]
Legend, I'll give it a bash as soon as I get a chance to, cheers!
